"Unable to determine the correct CMake Generator for Visual Studio"

I have VS17 installed, and now trying to build dotnet/winforms from cli fails due to the following error message:

```
C:\Users\user\.nuget\packages\microsoft.dotnet.cmake.sdk\6.0.0-beta.21076.1\build\Microsoft.DotNet.CMake.Sdk.targets(53,5): error : Unable to determine the correct CMake Generator for Visual Studio. The SDK may need to be updated to understand a new VS version. Unknown Visual Studio version 17.0.31222.107 [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj]
```

Naively I tried to change Microsoft.DotNet.CMake.Sdk.targets to:
```diff
    <PropertyGroup>
      <VSGenerator Condition="$(_HighestCompatibleVSVersion.StartsWith(16.))">Visual Studio 16 2019</VSGenerator>
+     <VSGenerator Condition="$(_HighestCompatibleVSVersion.StartsWith(17.))">Visual Studio 17 2020</VSGenerator>
    </PropertyGroup>
```

But this fails even more spectacularly:
```
EXEC : CMake error : Could not create named generator Visual Studio 17 2020 [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj]
  
  Generators
  * Visual Studio 16 2019        = Generates Visual Studio 2019 project files.
                                   Use -A option to specify architecture.
    Visual Studio 15 2017 [arch] = Generates Visual Studio 2017 project files.
                                   Optional [arch] can be "Win64" or "ARM".
    Visual Studio 14 2015 [arch] = Generates Visual Studio 2015 project files.
                                   Optional [arch] can be "Win64" or "ARM".
    Visual Studio 12 2013 [arch] = Generates Visual Studio 2013 project files.
                                   Optional [arch] can be "Win64" or "ARM".
    Visual Studio 11 2012 [arch] = Generates Visual Studio 2012 project files.
                                   Optional [arch] can be "Win64" or "ARM".
    Visual Studio 10 2010 [arch] = Generates Visual Studio 2010 project files.
                                   Optional [arch] can be "Win64" or "IA64".
    Visual Studio 9 2008 [arch]  = Generates Visual Studio 2008 project files.
                                   Optional [arch] can be "Win64" or "IA64".
    Borland Makefiles            = Generates Borland makefiles.
    NMake Makefiles              = Generates NMake makefiles.
    NMake Makefiles JOM          = Generates JOM makefiles.
    MSYS Makefiles               = Generates MSYS makefiles.
    MinGW Makefiles              = Generates a make file for use with
                                   mingw32-make.
    Unix Makefiles               = Generates standard UNIX makefiles.
    Green Hills MULTI            = Generates Green Hills MULTI files
                                   (experimental, work-in-progress).
    Ninja                        = Generates build.ninja files.
    Ninja Multi-Config           = Generates build-<Config>.ninja files.
    Watcom WMake                 = Generates Watcom WMake makefiles.
    CodeBlocks - MinGW Makefiles = Generates CodeBlocks project files.
    CodeBlocks - NMake Makefiles = Generates CodeBlocks project files.
    CodeBlocks - NMake Makefiles JOM
                                 = Generates CodeBlocks project files.
    CodeBlocks - Ninja           = Generates CodeBlocks project files.
    CodeBlocks - Unix Makefiles  = Generates CodeBlocks project files.
    CodeLite - MinGW Makefiles   = Generates CodeLite project files.
    CodeLite - NMake Makefiles   = Generates CodeLite project files.
    CodeLite - Ninja             = Generates CodeLite project files.
    CodeLite - Unix Makefiles    = Generates CodeLite project files.
    Sublime Text 2 - MinGW Makefiles
                                 = Generates Sublime Text 2 project files.
    Sublime Text 2 - NMake Makefiles
                                 = Generates Sublime Text 2 project files.
    Sublime Text 2 - Ninja       = Generates Sublime Text 2 project files.
    Sublime Text 2 - Unix Makefiles
                                 = Generates Sublime Text 2 project files.
    Kate - MinGW Makefiles       = Generates Kate project files.
    Kate - NMake Makefiles       = Generates Kate project files.
    Kate - Ninja                 = Generates Kate project files.
    Kate - Unix Makefiles        = Generates Kate project files.
    Eclipse CDT4 - NMake Makefiles
                                 = Generates Eclipse CDT 4.0 project files.
    Eclipse CDT4 - MinGW Makefiles
                                 = Generates Eclipse CDT 4.0 project files.
    Eclipse CDT4 - Ninja         = Generates Eclipse CDT 4.0 project files.
    Eclipse CDT4 - Unix Makefiles= Generates Eclipse CDT 4.0 project files.

C:\Users\user\.nuget\packages\microsoft.dotnet.cmake.sdk\6.0.0-beta.21076.1\build\Microsoft.DotNet.CMake.Sdk.targets(176,5): error MSB3073: The command " [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj] 
C:\Users\user\.nuget\packages\microsoft.dotnet.cmake.sdk\6.0.0-beta.21076.1\build\Microsoft.DotNet.CMake.Sdk.targets(176,5): error MSB3073:        [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj]        
C:\Users\user\.nuget\packages\microsoft.dotnet.cmake.sdk\6.0.0-beta.21076.1\build\Microsoft.DotNet.CMake.Sdk.targets(176,5): error MSB3073:        [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj]        
C:\Users\user\.nuget\packages\microsoft.dotnet.cmake.sdk\6.0.0-beta.21076.1\build\Microsoft.DotNet.CMake.Sdk.targets(176,5): error MSB3073:        cmake -G "Visual Studio 17 2020" -A x64 -B "C:\Development\winforms\artifacts\obj\NativeTests\x64\Debug" -S "C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests" -DCMAKE_INSTALL_PREFIX=C:\Development\winforms\artifacts\bin\NativeTests\x64\Debug [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj]
C:\Users\igveliko\.nuget\packages\microsoft.dotnet.cmake.sdk\6.0.0-beta.21076.1\build\Microsoft.DotNet.CMake.Sdk.targets(176,5): error MSB3073:       " exited with code 1. [C:\Development\winforms\src\System.Windows.Forms\tests\InteropTests\NativeTests\NativeTests.proj]
```

The follow change seems to work:
```diff
  <PropertyGroup Condition="$([MSBuild]::IsOsPlatform(Windows))">
    <VSWherePath>$([MSBuild]::NormalizePath('$(Pkgvswhere)','tools','vswhere.exe'))</VSWherePath>
    <_VSNativeCompilerComponentArchName Condition="'$(Platform)' == 'x86'">x86.x64</_VSNativeCompilerComponentArchName>
    <_VSNativeCompilerComponentArchName Condition="'$(Platform)' == 'x64'">x86.x64</_VSNativeCompilerComponentArchName>
    <_VSNativeCompilerComponentArchName Condition="'$(Platform)' == 'arm'">ARM</_VSNativeCompilerComponentArchName>
    <_VSNativeCompilerComponentArchName Condition="'$(Platform)' == 'arm64'">ARM64</_VSNativeCompilerComponentArchName>
    <_VSNativeCompilerComponentName>Microsoft.VisualStudio.Component.VC.Tools.$(_VSNativeCompilerComponentArchName)</_VSNativeCompilerComponentName>
+   <_VSTargetRange>"[16.0,16.99)"</_VSTargetRange>
  </PropertyGroup>

    <Exec
-     Command="$(VSWherePath) -latest -prerelease -products * -requires $(_VSNativeCompilerComponentName) -property installationVersion"
+     Command="$(VSWherePath) -latest -prerelease -version $(_VSTargetRange) -products * -requires $(_VSNativeCompilerComponentName) -property installationVersion"
      EchoOff="true"
      ConsoleToMsBuild="true"
      StandardOutputImportance="Low">
      <Output TaskParameter="ConsoleOutput" PropertyName="_HighestCompatibleVSVersion" />
    </Exec>
```

--- a/src/Microsoft.DotNet.CMake.Sdk/build/Microsoft.DotNet.CMake.Sdk.targets
+++ b/src/Microsoft.DotNet.CMake.Sdk/build/Microsoft.DotNet.CMake.Sdk.targets
@@ -30,6 +30,8 @@
     <PropertyGroup>
       <_CMakeMultiConfigurationGenerator>true</_CMakeMultiConfigurationGenerator>
       <_CMakePassArchitectureToGenerator>true</_CMakePassArchitectureToGenerator>
+      <!-- We limit the VS version range for the generator since CMake only knows about specific VS versions -->
+      <VSGeneratorVersionRange Condition="'$(VSGeneratorVersionRange)' == ''">[,17.0)</VSGeneratorVersionRange>
       <!-- Visual Studio uses the Win32 name for the x86 target architecture. -->
       <Platform Condition="'$(Platform)' == 'x86'">Win32</Platform>
     </PropertyGroup>
@@ -37,7 +39,7 @@
     <Error Condition="'$(_VSNativeCompilerComponentArchName)' == ''" Text="Unable to determine if Visual Studio has MSVC tools installed for the '$(Platform)' architecture." />
 
     <Exec
-      Command="$(VSWherePath) -latest -prerelease -products * -requires $(_VSNativeCompilerComponentName) -property installationVersion"
+      Command="$(VSWherePath) -latest -prerelease -products * -version $(VSGeneratorVersionRange) -requires $(_VSNativeCompilerComponentName) -property installationVersion"
       EchoOff="true"
       ConsoleToMsBuild="true"
       StandardOutputImportance="Low">

